### ISSUE TITLE
How to normalize a tensor to [-1,1] in tensorflow???

### ISSUE THREAD

**Issue by longmalongma** [qway/nerfmeshes#1]
What is the function of your function? Is this to normalize the tensor to lie between [-1, 1]?How to normalize a tensor to [-1,1] in tensorflow???

frequency_bands = torch.linspace(
            2.0 ** 0.0,
            2.0 ** (num_encoding_functions - 1),
            num_encoding_functions,
            dtype=tensor.dtype,
            device=tensor.device,
        )

**Comment by DomainFlag** [comment]
@longmalongma This is used for positional encoding, it linearly samples between the lowest and highest frequency (octave). When applying cos, sin it helps draw manifold representation of the same 3D position input (x, y, z) that is scaled by those encodings allowing the network to learn better high-frequency functions (captures better the detail). 

https://github.com/qway/nerfmeshes/blob/8ec9383b6af30b5160d77984a40fc56d1759e13c/src/nerf/nerf_helpers.py#L142-L157

The snippet you provided is not actually logarithmic sampling (1, 2, 4, 8...) with the assumption that high-frequency functions are not totally independent thus there are some crossings especially for those low-frequency bins containing multiple samples compared to the highest ones.

**Comment by DomainFlag** [comment]
@longmalongma The repo isn't in the final state, as we are working on refactoring the existent PyTorch lightning implementation and finish by one week merging the experimental branches. 

With this repo you will have a modern and very efficient re-implementation of the official NeRF paper in PyTorch Lightning, mesh exporting with appearance based on 2nd query guided by the normals, self managing tree of sparse neural voxel fields that can be ray-cased for efficient sampling -> no need for 2nd finer model resulting in twice as fast training time.

If you are interested, just hold on one more week and comeback then + there will be actual documentation to help you through.

**Comment by longmalongma** [comment]
> @longmalongma The repo isn't in the final state, as we are working on refactoring the existent PyTorch lightning implementation and finish by one week merging the experimental branches.
> 
> With this repo you will have a modern and very efficient re-implementation of the official NeRF paper in PyTorch Lightning, mesh exporting with appearance based on 2nd query guided by the normals, self managing tree of sparse neural voxel fields that can be ray-cased for efficient sampling -> no need for 2nd finer model resulting in twice as fast training time.
> 
> If you are interested, just hold on one more week and comeback then + there will be actual documentation to help you through.

Thanks, To realize positional encoding, how to write the code based on TensorFlow?

**Comment by DomainFlag** [comment]
@longmalongma this is not the right place to answer those kind of questions, there are the likes of stackoverflow that are designed for that. But you can start from here [https://www.tensorflow.org/api_docs/python/tf/linspace]( https://www.tensorflow.org/api_docs/python/tf/linspace)

**Comment by longmalongma** [comment]
> @longmalongma this is not the right place to answer those kind of questions, there are the likes of stackoverflow that are designed for that. But you can start from here https://www.tensorflow.org/api_docs/python/tf/linspace

Can I use the function 'inputs_normal = inputs/tf.reduce_max(tf.abs(inputs))' to instead of your'torch.linspace() '? Because they can all normalize tensor to [-1,1].
